File: backend/models/PowModel.py
import sys
from .db import db

from sqlalchemy_utils import JSONType, ChoiceType
from .TermModel import Term
from .CourseModel import Course
from .SpecializationModel import Specialization


STATUS = [
    ('Void', 'Void'),
    ('Active', 'Active'),
    ('Pending Approval', 'Pending Approval'),
    ('POW Approved', 'POW Approved'),
    ('POW Rejected', 'POW Rejected'),
    ('Pending Graduation', 'Pending Graduation'),
    ('Graduation Approved', 'Graduation Approved'),
    ('Graduation Rejected', 'Graduation Rejected'),
    ('Graduated', 'Graduated'),
]


class POW(db.Model):
    id = db.Column(db.Integer, primary_key=True)
    student_id = db.Column(db.Integer, db.ForeignKey('user_role.id'), nullable=False)
    creation_date = db.Column(db.DateTime)
    petition_date = db.Column(db.DateTime)
    candidacy_approved_by_id = db.Column(db.Integer, db.ForeignKey('user_role.id'))
    candidacy_approval_date = db.Column(db.DateTime)
    graduation_approved_by_id = db.Column(db.Integer, db.ForeignKey('user_role.id'))
    graduation_approval_date = db.Column(db.DateTime)
    specialization_id = db.Column(db.Integer, db.ForeignKey('specialization.id'), nullable=False)
    orientation_term_id = db.Column(db.Integer, db.ForeignKey('term.id'))
    first_class_term_id = db.Column(db.Integer, db.ForeignKey('term.id'))
    status = db.Column(ChoiceType(STATUS), default='Void')
    pow_courses = db.relationship('StudentPOWCourse', backref='pow', cascade="all, delete-orphan")
    completion_date = db.Column(db.DateTime)
    orientation_term = db.relationship('Term', backref='orientation_pows', foreign_keys=[orientation_term_id])
    first_class_term = db.relationship('Term', backref='first_class_pows', foreign_keys=[first_class_term_id])
    specialization = db.relationship('Specialization', backref='pow')
    student = db.relationship('UserRole', backref='pow_student', foreign_keys=[student_id])
    candidacy_approved_by = db.relationship('UserRole', backref='pow_candidacy_approved_by',
                                            foreign_keys=[candidacy_approved_by_id])
    graduation_approved_by = db.relationship('UserRole', backref='pow_graduation_approved_by',
                                             foreign_keys=[graduation_approved_by_id])


class StudentPOWCourse(db.Model):
    id = db.Column(db.Integer, primary_key=True)
    course_id = db.Column(db.Integer, db.ForeignKey('course.id'), nullable=False)
    pow_id = db.Column(db.Integer, db.ForeignKey('POW.id'), nullable=False)
    term_id = db.Column(db.Integer, db.ForeignKey('term.id'))
    course_type = db.Column(db.String(80))
    course = db.relationship('Course', backref='pow_course')
    term = db.relationship('Term', backref='pow_course')

File: frontend/src/app/shared/models/pow.model.ts
import { CourseModel } from './course.model';
import { UserRoleModel } from '../../core/models/user-role.model';
import { TermModel } from './term.model';
import { SpecializationModel } from './specialization.model';
import { CompletedCourseModel } from './completed-course.model';

export interface NewPowCourseInterface {
  course_id?: number;
  term_id?: number | null;
  course_type?: string;
  credits?: any;
}

type PowStatus =
  | 'Void'
  | 'Active'
  | 'Pending Approval'
  | 'POW Approved'
  | 'POW Rejected'
  | 'Pending Graduation'
  | 'Graduation Approved'
  | 'Graduation Rejected'
  | 'Graduated';

export class PowModel {
  id: number | undefined;
  student: UserRoleModel | undefined;
  creationDate: Date | undefined;
  petitionDate: Date | undefined;
  candidacyApproval: UserRoleModel | undefined;
  candidacyApprovalDate: Date | undefined;
  graduationApproval: UserRoleModel | undefined;
  graduationApprovalDate: Date | undefined;
  status: PowStatus | undefined;
  powCourses: PowCourseModel[] | undefined;
  specialization: SpecializationModel | undefined;
  orientationTerm: TermModel | undefined;
  firstClassTerm: TermModel | undefined;
  completionDate: Date | undefined;
  completedCourses: CompletedCourseModel[] | undefined;

  constructor(powModel: { [key: string]: any } | undefined) {
    this.id = powModel?.id;
    this.student = new UserRoleModel({ ...powModel?.student });
    this.petitionDate =
      powModel?.petition_date != null
        ? new Date(powModel?.petition_date)
        : undefined;
    this.creationDate =
      powModel?.creation_date != null
        ? new Date(powModel?.creation_date)
        : undefined;
    this.candidacyApproval = new UserRoleModel({
      ...powModel?.candidacy_approved_by,
    });
    this.candidacyApprovalDate =
      powModel?.candidacy_approval_date != null
        ? new Date(powModel?.candidacy_approval_date)
        : undefined;
    this.specialization = new SpecializationModel({
      ...powModel?.specialization,
    });
    this.graduationApproval = new UserRoleModel({
      ...powModel?.graduation_approved_by,
    });
    this.graduationApprovalDate =
      powModel?.graduation_approval_date != null
        ? new Date(powModel?.graduation_approval_date)
        : undefined;
    this.powCourses = powModel?.pow_courses?.map(
      (course: { [key: string]: any } | undefined) =>
        new PowCourseModel({ ...course })
    );
    this.completedCourses = powModel?.completed_courses?.map(
      (course: { [key: string]: any } | undefined) =>
        new CompletedCourseModel({ ...course })
    );
    this.orientationTerm = new TermModel({ ...powModel?.orientation_term });
    this.firstClassTerm = new TermModel({ ...powModel?.first_class_term });
    this.completionDate =
      powModel?.completion_date != null
        ? new Date(powModel?.completion_date)
        : undefined;
    this.status = powModel?.status;
  }
}

export class PowCourseModel {
  id: number | undefined;
  course: CourseModel | undefined;
  pow: PowModel | undefined;
  term: TermModel | undefined;
  courseType: string | undefined;

  constructor(powCourseModel: { [key: string]: any } | undefined) {
    this.id = powCourseModel?.id;
    this.course = new CourseModel({ ...powCourseModel?.course });
    this.pow = new PowModel({ ...powCourseModel?.pow });
    this.term = new TermModel({ ...powCourseModel?.term });
    this.courseType = powCourseModel?.course_type;
  }
}
